src/main/kotlin/com/ampnet/mailservice/service/impl/FileServiceImpl.kt

```diff
@@ -2,41 +2,45 @@ package com.ampnet.mailservice.service.impl
 
 import com.ampnet.mailservice.exception.ResourceNotFoundException
 import com.ampnet.mailservice.service.FileService
-import org.springframework.core.io.Resource
-import org.springframework.http.HttpEntity
-import org.springframework.http.HttpMethod
-import org.springframework.http.ResponseEntity
 import org.springframework.stereotype.Service
-import org.springframework.web.client.RestClientException
-import org.springframework.web.client.RestTemplate
-import org.springframework.web.client.exchange
+import java.io.ByteArrayOutputStream
 import java.io.IOException
-import java.io.InputStream
+import java.net.HttpURLConnection
+import java.net.MalformedURLException
+import java.net.URL
 
 @Service
-class FileServiceImpl(private val restTemplate: RestTemplate) : FileService {
+class FileServiceImpl : FileService {
 
-    private val termsOfService: MutableMap<String, InputStream> = mutableMapOf()
+    private val termsOfService: MutableMap<String, ByteArray> = mutableMapOf()
 
     @Throws(ResourceNotFoundException::class)
-    override fun getTermsOfService(url: String): InputStream {
+    override fun getTermsOfService(url: String): ByteArray {
         termsOfService[url]?.let { return it }
+        val byteArray = getFileContent(url)
+        termsOfService[url] = byteArray
+        return byteArray
+    }
+
+    private fun getFileContent(url: String): ByteArray {
         try {
-            val responseEntity =
-                restTemplate.exchange<Resource>(url, HttpMethod.GET, HttpEntity.EMPTY)
-            val inputStream = readInputStream(responseEntity, url)
-            termsOfService[url] = inputStream
-            return inputStream
-        } catch (ex: RestClientException) {
-            throw ResourceNotFoundException("Error while reading response from $url", ex)
+            val connection = getURLFromString(url).openConnection() as HttpURLConnection
+            val inputStream = connection.inputStream
+            val outputStream = ByteArrayOutputStream()
+            inputStream.copyTo(outputStream)
+            inputStream.close()
+            outputStream.close()
+            return outputStream.toByteArray()
```

Comment:
There is always a problem with close functions in try-catch block. On exception, there is a possibility that the stream will not be closed. Usually, it is done in finally block, but Kotlin has a function: `use` https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/use.html.

```suggestion
            val connection = getURLFromString(url).openConnection() as HttpURLConnection
            connection.inputStream.use { input ->
                ByteArrayOutputStream().use { output ->
                    input.copyTo(output)
                    return output.toByteArray()
                }
            }
```

This is my first hint, check if there is a better solution.

Reply:
It seems to be the prefered way to close a resource in Kotlin. There are suggestions to avoid nesting of `use` blocks by using an extension function, but in our case it's not problematic.